docs/design/001-prometheus.md

```diff
@@ -0,0 +1,447 @@
+
+The implementation is based on the existing parser that uses reflection underneath, a PoC is
+implemented in
+[https://github.com/inspektor-gadget/inspektor-gadget/tree/mauricio/experiments/prometheus](https://github.com/inspektor-gadget/inspektor-gadget/tree/mauricio/experiments/prometheus)
+
+### Collection in eBPF
+
+We should extend the gadgets to collect some common metrics in eBPF to make this solution more
+performant. This should be the preferred way of collecting metrics, even if it isn't as flexible as
+the implementation using the events (and it's harder to implement). In this case, we'll have to define a
+list of common metrics that are exposed by each gadget (TODO).
+
+Gadgets' eBPF code would require two new constants:
+
+```c
+bool enable_events
+bool enable_metrics
+```
+
+When metric collection is enabled, it expects maps that it can fill. The gadget/operator will provide such
+maps with a layout depending on the users' configuration.
+
+The biggest issue here is that we basically need to have counters for each of the possible tuples of
+the requested labels (which are dynamic). This would be solved by a BPF_HASH_MAP (potentially
+PER_CPU) with the key looking like this for example:
+
+```c
+struct metric_key_t {
+__u64 mount_ns_id;
+__u32 reason;
+}
+```
+
+With each added label, the key length increases. Adding for example the SADDR, the key would look
+like this:
+
+```c
+struct metric_key_t {
+__u64 mount_ns_id;
+__u8 saddr[16];
+__u32 reason;
+}
+```
+
+The actual value consists of a simpler struct, containing for example just a counter variable,
+buckets for histograms, etc. (and maybe a last-access timestamp).
+
+The operator would then periodically iterate over the map and update the exported metrics.
+
+We'd have to think about pruning the maps periodically as well, otherwise the maps would only grow -
+the mentioned timestamp could help with that. We also have to notify the user about possible
+overflows.
+
+#### Use of Macros
+
+To be able to use maps like described above, the keys of the maps would have to be sized
+dynamically. This can be achieved by using macros and consts that switch on/off certain fields on
+demand (I've started work on a PoC).
+
+A definition like this
+
+```c
+#define METRIC_LIST_KEYS(X) \
+ X(MntNS, __u64, 8) \
+ X(Reason, __u32, 4)
+```
+
+could then create required functions (like offset helpers) and volatile consts that will be set upon
+starting the gadget.
+
+# Compatibility with script and BYOB (bring your own bpf) gadgets
+
+These two gadgets allow the user to inject custom eBPF programs. It should be possible to
+also support metric collection using this solution. The structure of the eBPF maps that collect the
+metrics should be well defined to create a contract between the eBPF programs and the
+Operator/Gadget in user space.
+
+## BYOB
+
+We have to document what is the structure of the maps and the user will be responsible for creating
+and filling them in their eBPF programs
```

Comment:
Another possibility is using a bpf extension defined by Inspektor Gadget. Then the byob code would call the bpf extension:
```
struct event event = {};
event.field1 = ...;
ig_count_event(ctx, event);
```

The bpf extension would be built dynamically after inspecting the btf type of `struct event`.

That would reduce the surface of the contract between the eBPF programs and the Operator/Gadget in user space.

But we can explore that in a future PR.